Backslash at the end of map-entry-value string cannot be unmarshaled

Seems that this was a lucky catch because if you have another character at the end e.g. `{"c":"\\a"}` it works perfectly.

```
package main

import (
	"encoding/json"
	"fmt"
	"os"

	gojson "github.com/goccy/go-json"
	"github.com/kr/pretty"
)

type SomeStruct struct {
	Data []byte
}

func main() {
	original := []byte(`{"c":"\\"}`)

	validEncodingJSON := json.Valid(original)
	validGoJSON := gojson.Valid(original)

	fmt.Printf("encoding/json = %v vs go/json = %v\n", validEncodingJSON, validGoJSON)
	if validEncodingJSON != validGoJSON {
		fmt.Println("ERROR: Unequal \"Valid\" results")

		os.Exit(1)
	}

	var resultEncodingJSON map[string]json.RawMessage
	if err := json.Unmarshal(original, &resultEncodingJSON); err != nil {
		panic(err)
	}
	resultEncodingJSONString := readable(resultEncodingJSON)
	fmt.Println("encoding/json = \n", resultEncodingJSONString)

	var resultGoJSON map[string]gojson.RawMessage
	if err := gojson.Unmarshal(original, &resultGoJSON); err != nil {
		panic(err)
	}
	resultGoJSONString := readable(resultGoJSON)
	fmt.Println("go/json = \n", resultGoJSONString)

	if resultEncodingJSONString != resultGoJSONString {
		fmt.Println("ERROR: Unequal \"Unmarshal\" results")

		os.Exit(1)
	}
}

func readable(object interface{}) string {
	return fmt.Sprintf("%# v", pretty.LazyFormatter(object))
}
```

results in

```
encoding/json = true vs go/json = true
encoding/json = 
 map[string][]uint8{
    "c": {0x22, 0x5c, 0x5c, 0x22} ("\"\\\\\""),
}
panic: json: string of object unexpected end of JSON input

goroutine 1 [running]:
main.main()
        /home/zimmski/symflower/tt.go:38 +0x51c
exit status 2

--- a/decode_context.go
+++ b/decode_context.go
@@ -55,10 +55,12 @@ func skipObject(buf []byte, cursor, depth int64) (int64, error) {
 			for {
 				cursor++
 				switch buf[cursor] {
-				case '"':
-					if buf[cursor-1] == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if buf[cursor] == nul {
+						return 0, errUnexpectedEndOfJSON("string of object", cursor)
 					}
+				case '"':
 					goto SWITCH_OUT
 				case nul:
 					return 0, errUnexpectedEndOfJSON("string of object", cursor)
@@ -99,10 +101,12 @@ func skipArray(buf []byte, cursor, depth int64) (int64, error) {
 			for {
 				cursor++
 				switch buf[cursor] {
-				case '"':
-					if buf[cursor-1] == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if buf[cursor] == nul {
+						return 0, errUnexpectedEndOfJSON("string of object", cursor)
 					}
+				case '"':
 					goto SWITCH_OUT
 				case nul:
 					return 0, errUnexpectedEndOfJSON("string of object", cursor)
@@ -130,10 +134,12 @@ func skipValue(buf []byte, cursor, depth int64) (int64, error) {
 			for {
 				cursor++
 				switch buf[cursor] {
-				case '"':
-					if buf[cursor-1] == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if buf[cursor] == nul {
+						return 0, errUnexpectedEndOfJSON("string of object", cursor)
 					}
+				case '"':
 					return cursor + 1, nil
 				case nul:
 					return 0, errUnexpectedEndOfJSON("string of object", cursor)
@@ -184,18 +190,8 @@ func skipValue(buf []byte, cursor, depth int64) (int64, error) {
 			cursor += 5
 			return cursor, nil
 		case 'n':
-			buflen := int64(len(buf))
-			if cursor+3 >= buflen {
-				return 0, errUnexpectedEndOfJSON("null", cursor)
-			}
-			if buf[cursor+1] != 'u' {
-				return 0, errUnexpectedEndOfJSON("null", cursor)
-			}
-			if buf[cursor+2] != 'l' {
-				return 0, errUnexpectedEndOfJSON("null", cursor)
-			}
-			if buf[cursor+3] != 'l' {
-				return 0, errUnexpectedEndOfJSON("null", cursor)
+			if err := validateNull(buf, cursor); err != nil {
+				return 0, err
 			}
 			cursor += 4
 			return cursor, nil

--- a/decode_stream.go
+++ b/decode_stream.go
@@ -126,10 +126,18 @@ func (s *stream) skipObject(depth int64) error {
 			for {
 				cursor++
 				switch char(p, cursor) {
-				case '"':
-					if char(p, cursor-1) == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if char(p, cursor) == nul {
+						s.cursor = cursor
+						if s.read() {
+							s.cursor-- // for retry current character
+							_, cursor, p = s.stat()
+							continue
+						}
+						return errUnexpectedEndOfJSON("string of object", cursor)
 					}
+				case '"':
 					goto SWITCH_OUT
 				case nul:
 					s.cursor = cursor
@@ -183,10 +191,18 @@ func (s *stream) skipArray(depth int64) error {
 			for {
 				cursor++
 				switch char(p, cursor) {
-				case '"':
-					if char(p, cursor-1) == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if char(p, cursor) == nul {
+						s.cursor = cursor
+						if s.read() {
+							s.cursor-- // for retry current character
+							_, cursor, p = s.stat()
+							continue
+						}
+						return errUnexpectedEndOfJSON("string of object", cursor)
 					}
+				case '"':
 					goto SWITCH_OUT
 				case nul:
 					s.cursor = cursor
@@ -235,10 +251,18 @@ func (s *stream) skipValue(depth int64) error {
 			for {
 				cursor++
 				switch char(p, cursor) {
-				case '"':
-					if char(p, cursor-1) == '\\' {
-						continue
+				case '\\':
+					cursor++
+					if char(p, cursor) == nul {
+						s.cursor = cursor
+						if s.read() {
+							s.cursor-- // for retry current character
+							_, cursor, p = s.stat()
+							continue
+						}
+						return errUnexpectedEndOfJSON("value of string", s.totalOffset())
 					}
+				case '"':
 					s.cursor = cursor + 1
 					return nil
 				case nul:

--- a/decode_test.go
+++ b/decode_test.go
@@ -3417,3 +3417,88 @@ func TestDecodeByteSliceNull(t *testing.T) {
 		}
 	})
 }
+
+func TestDecodeBackSlash(t *testing.T) {
+	t.Run("unmarshal", func(t *testing.T) {
+		t.Run("string", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.Unmarshal([]byte(`{"c":"\\"}`), &v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.Unmarshal([]byte(`{"c":"\\"}`), &v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+		t.Run("array", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.Unmarshal([]byte(`{"c":["\\"]}`), &v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.Unmarshal([]byte(`{"c":["\\"]}`), &v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+		t.Run("object", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.Unmarshal([]byte(`{"c":{"\\":"\\"}}`), &v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.Unmarshal([]byte(`{"c":{"\\":"\\"}}`), &v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+	})
+	t.Run("stream", func(t *testing.T) {
+		t.Run("string", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.NewDecoder(strings.NewReader(`{"c":"\\"}`)).Decode(&v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.NewDecoder(strings.NewReader(`{"c":"\\"}`)).Decode(&v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+		t.Run("array", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.NewDecoder(strings.NewReader(`{"c":["\\"]}`)).Decode(&v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.NewDecoder(strings.NewReader(`{"c":["\\"]}`)).Decode(&v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+		t.Run("object", func(t *testing.T) {
+			var v1 map[string]stdjson.RawMessage
+			if err := stdjson.NewDecoder(strings.NewReader(`{"c":{"\\":"\\"}}`)).Decode(&v1); err != nil {
+				t.Fatal(err)
+			}
+			var v2 map[string]json.RawMessage
+			if err := json.NewDecoder(strings.NewReader(`{"c":{"\\":"\\"}}`)).Decode(&v2); err != nil {
+				t.Fatal(err)
+			}
+			if len(v1) != len(v2) || !bytes.Equal(v1["c"], v2["c"]) {
+				t.Fatalf("failed to decode backslash: expected %#v but got %#v", v1, v2)
+			}
+		})
+	})
+}

